### PR TITLE
extract child mesh from tag

### PR DESCRIPTION
**Pull request by JcDai**
add generic functions to extract a child mesh from a provided long-type tag (or its handle), and register the extracted child mesh to the parent mesh. (The child mesh should not break the orientation of the parent mesh)

- [x] extract child mesh by tag
- [x] extract child mesh by tag handle
- [x] generalize the functions
- [x] check if the orientation keeps in the childmesh

### CODE REVIEW

**Review by JcDai — COMMENTED** [review]
need to check if the current implementation of filling the child mesh matrix preserve the orientation for all possible cases.

File: src/wmtk/Mesh.hpp [hunk 1]
```diff
@@ -62,6 +63,17 @@ class Mesh : public std::enable_shared_from_this<Mesh>
...
+    friend std::shared_ptr<Mesh> multimesh::utils::extract_and_register_child_mesh_from_tag(
+        Mesh& m,
+        const std::string& tag,
+        const long tag_value,
+        const PrimitiveType& pt);
+    friend std::shared_ptr<Mesh> multimesh::utils::extract_and_register_child_mesh_from_tag_handle(
+        Mesh& m,
+        const MeshAttributeHandle<long>& tag_handle,
+        const long tag_value,
+        const PrimitiveType& pt);
```
Comment (mtao) [inline src/wmtk/Mesh.hpp:75]: this pt is redundant with tag_handle.primitive_type() isn't it?

File: src/wmtk/multimesh/utils/extract_child_mesh_from_tag.cpp [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    case PrimitiveType::Vertex: throw("not implemented");
+    case PrimitiveType::Edge: {
+        std::map<long, long> parent_to_child_vertex_map;
+        long child_vertex_count = 0;
+
+        RowVectors2l edge_mesh_matrix;
+        edge_mesh_matrix.resize(tagged_tuples.size(), 2);
+
+        for (long i = 0; i < tagged_tuples.size(); ++i) {
+            const long v1 = m.id(tagged_tuples[i], PrimitiveType::Vertex);
```
Comment (mtao) [inline src/wmtk/multimesh/utils/extract_child_mesh_from_tag.cpp:51]: make this a `const std::array<long,2>`. this allows for line 55 and 59 to turn into a for loop, same with 65/66.

File: src/wmtk/multimesh/utils/extract_child_mesh_from_tag.cpp [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        return child_ptr;
+    }
+    case PrimitiveType::Face: {
+        std::map<long, long> parent_to_child_vertex_map;
+        long child_vertex_count = 0;
+
+        RowVectors3l tri_mesh_matrix;
+        tri_mesh_matrix.resize(tagged_tuples.size(), 3);
+        for (long i = 0; i < tagged_tuples.size(); ++i) {
+            // TODO: check if this break the orientation of the map
```
Comment (mtao) [inline src/wmtk/multimesh/utils/extract_child_mesh_from_tag.cpp:92]: same as before about using an array

File: src/wmtk/multimesh/utils/extract_child_mesh_from_tag.cpp [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+        RowVectors2l edge_mesh_matrix;
+        edge_mesh_matrix.resize(tagged_tuples.size(), 2);
+
+        for (long i = 0; i < tagged_tuples.size(); ++i) {
+            const long v1 = m.id(tagged_tuples[i], PrimitiveType::Vertex);
+            const long v2 = m.id(m.switch_vertex(tagged_tuples[i]), PrimitiveType::Vertex);
+
+            // check and add v1, v2 to the vertex map
+            if (parent_to_child_vertex_map.find(v1) == parent_to_child_vertex_map.end()) {
```
Comment (mtao) [inline src/wmtk/multimesh/utils/extract_child_mesh_from_tag.cpp:55]: there's a pattern like
```
size_t size = parent_to_Child_Vertex_map.size();
parent_to_child_vertex_map.try_emplace(v,size);
```
that is a little easier to read i think

File: src/wmtk/Mesh.hpp [hunk 5]
```diff
@@ -62,6 +63,17 @@ class Mesh : public std::enable_shared_from_this<Mesh>
     friend class multimesh::MultiMeshVisitorExecutor;
     friend class operations::utils::UpdateEdgeOperationMultiMeshMapFunctor;
 
+    friend std::shared_ptr<Mesh> multimesh::utils::extract_and_register_child_mesh_from_tag(
+        Mesh& m,
```
Comment (mtao) [inline src/wmtk/Mesh.hpp:67]: this done doesn't need to be a friend

**Review by mtao — CHANGES_REQUESTED** [review]
mostly there. just deal with my comments

**Review by mtao — APPROVED** [review]
lgtm

File: src/wmtk/multimesh/utils/extract_child_mesh_from_tag.hpp [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+ * @param tag tag of type long
+ * @param tag_value target tag value
+ * @param pt primitive type of the tag
+ * @return std::shared_ptr<Mesh> the shared pointer to the child mesh
+ */
+std::shared_ptr<Mesh> extract_and_register_child_mesh_from_tag(
+    Mesh& m,
+    const std::string& tag,
+    const long tag_value,
+    const PrimitiveType& pt);
```
Comment (daniel-zint) [inline src/wmtk/multimesh/utils/extract_child_mesh_from_tag.hpp:32]: `PrimitiveType` should not be passed by reference. Just pass it by value.

File: src/wmtk/multimesh/utils/extract_child_mesh_from_tag.cpp [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+
+std::shared_ptr<Mesh> extract_and_register_child_mesh_from_tag_handle(
+    Mesh& m,
+    const MeshAttributeHandle<long>& tag_handle,
+    const long tag_value)
+{
+    auto tags = m.create_const_accessor(tag_handle);
+    PrimitiveType pt = tag_handle.primitive_type();
```
Comment (daniel-zint) [inline src/wmtk/multimesh/utils/extract_child_mesh_from_tag.cpp:32]: `const`

File: src/wmtk/multimesh/utils/extract_child_mesh_from_tag.cpp [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+    std::vector<Tuple> tagged_tuples;
+    for (auto t : m.get_all(pt)) {
```
Comment (daniel-zint) [inline src/wmtk/multimesh/utils/extract_child_mesh_from_tag.cpp:35]: `const Tuple& t`

File: src/wmtk/multimesh/utils/extract_child_mesh_from_tag.cpp [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        if (tags.const_scalar_attribute(t) == tag_value) {
+            tagged_tuples.emplace_back(t);
+        }
+    }
+
+    switch (pt) {
+    case PrimitiveType::Vertex: throw("not implemented");
```
Comment (daniel-zint) [inline src/wmtk/multimesh/utils/extract_child_mesh_from_tag.cpp:42]: You can add a simple test case for that using `CHECK_THROWS()`. This test case can then later be modified when there is something actually happening.

File: src/wmtk/multimesh/utils/extract_child_mesh_from_tag.cpp [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    case PrimitiveType::Edge: {
+        std::map<long, long> parent_to_child_vertex_map;
+        // long child_vertex_count = 0;
+
+        RowVectors2l edge_mesh_matrix;
+        edge_mesh_matrix.resize(tagged_tuples.size(), 2);
+
+        for (long i = 0; i < tagged_tuples.size(); ++i) {
```
Comment (daniel-zint) [inline src/wmtk/multimesh/utils/extract_child_mesh_from_tag.cpp:50]: should be `size_t`

File: src/wmtk/multimesh/utils/extract_child_mesh_from_tag.cpp [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+            const std::array<long, 2> vs = {
+                {m.id(tagged_tuples[i], PrimitiveType::Vertex),
+                 m.id(m.switch_vertex(tagged_tuples[i]), PrimitiveType::Vertex)}};
+
+
+            // check and add v0, v1 to the vertex map
+            for (int k = 0; k < 2; k++) {
```
Comment (daniel-zint) [inline src/wmtk/multimesh/utils/extract_child_mesh_from_tag.cpp:57]: `size_t`

File: src/wmtk/multimesh/utils/extract_child_mesh_from_tag.cpp [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+        }
+
+        std::shared_ptr<EdgeMesh> child_ptr = std::make_shared<EdgeMesh>();
+        auto& child = *child_ptr;
+        child.initialize(edge_mesh_matrix);
+
+        std::vector<std::array<Tuple, 2>> child_to_parent_map(tagged_tuples.size());
+        assert(tagged_tuples.size() == child.capacity(PrimitiveType::Edge));
+
+        auto edgemesh_edge_tuples = child.get_all(PrimitiveType::Edge);
```
Comment (daniel-zint) [inline src/wmtk/multimesh/utils/extract_child_mesh_from_tag.cpp:71]: `const`

File: src/wmtk/multimesh/utils/extract_child_mesh_from_tag.cpp [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+
+        for (long i = 0; i < tagged_tuples.size(); ++i) {
```
Comment (daniel-zint) [inline src/wmtk/multimesh/utils/extract_child_mesh_from_tag.cpp:73]: `size_t`

File: src/wmtk/multimesh/utils/extract_child_mesh_from_tag.cpp [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+            child_to_parent_map[i] = {{edgemesh_edge_tuples[i], tagged_tuples[i]}};
+        }
+
+        m.register_child_mesh(child_ptr, child_to_parent_map);
+        return child_ptr;
+    }
+    case PrimitiveType::Face: {
```
Comment (daniel-zint) [inline src/wmtk/multimesh/utils/extract_child_mesh_from_tag.cpp:80]: Just looking at what Codecov reports, it seems like you did not test that part of the code. Please add a test.

File: tests/test_extract_child_mesh.cpp [hunk 15]
```diff
@@ -0,0 +1,589 @@
...
+using TMOE = decltype(std::declval<DEBUG_TriMesh>().get_tmoe(
+    wmtk::Tuple(),
+    std::declval<Accessor<long>&>()));
+
+constexpr PrimitiveType PV = PrimitiveType::Vertex;
+constexpr PrimitiveType PE = PrimitiveType::Edge;
+constexpr PrimitiveType PF = PrimitiveType::Face;
+
+
+namespace {} // namespace
```
Comment (daniel-zint) [inline tests/test_extract_child_mesh.cpp:26]: ?

File: tests/test_extract_child_mesh.cpp [hunk 16]
```diff
@@ -0,0 +1,589 @@
...
+#include "tools/DEBUG_Tuple.hpp"
+#include "tools/TriMesh_examples.hpp"
+
+using namespace wmtk;
+using namespace wmtk::tests;
+
+using TM = TriMesh;
+using TMOE = decltype(std::declval<DEBUG_TriMesh>().get_tmoe(
+    wmtk::Tuple(),
+    std::declval<Accessor<long>&>()));
```
Comment (daniel-zint) [inline tests/test_extract_child_mesh.cpp:19]: If I see it correctly, you are neither using `TM` nor `TMOE`. Please remove that.

File: tests/test_extract_child_mesh.cpp [hunk 17]
```diff
@@ -0,0 +1,589 @@
...
+
+TEST_CASE("test_extract_child_mesh", "[multimesh][2D]")
+{
+    SECTION("single_triangle")
+    {
+        DEBUG_TriMesh parent = single_triangle();
+        auto tag_handle = parent.register_attribute<long>("is_child", wmtk::PrimitiveType::Edge, 1);
+        auto tag_accessor = parent.create_accessor(tag_handle);
+
+        for (auto t : parent.get_all(PE)) {
```
Comment (daniel-zint) [inline tests/test_extract_child_mesh.cpp:37]: `const Tuple& t`

File: tests/test_extract_child_mesh.cpp [hunk 18]
```diff
@@ -0,0 +1,589 @@
...
+                PE);
+
+
+        const auto& p_mul_manager = parent.multi_mesh_manager();
+        REQUIRE(p_mul_manager.children().size() == 3);
+
+        const auto& child0 = *(p_mul_manager.children()[0].mesh);
+        const auto& child1 = *(p_mul_manager.children()[1].mesh);
+        const auto& child2 = *(p_mul_manager.children()[2].mesh);
+        REQUIRE(child0.get_all(PE).size() == 3);
```
Comment (daniel-zint) [inline tests/test_extract_child_mesh.cpp:104]: should be `CHECK`s

File: tests/test_extract_child_mesh.cpp [hunk 19]
```diff
@@ -0,0 +1,589 @@
...
+        REQUIRE(child0.get_all(PV).size() == 3);
+        REQUIRE(child1.get_all(PE).size() == 2);
+        REQUIRE(child1.get_all(PV).size() == 3);
+        REQUIRE(child2.get_all(PE).size() == 2);
+        REQUIRE(child2.get_all(PV).size() == 3);
+    }
+}
+
+//     auto& child0 = *child0_ptr;
```
Comment (daniel-zint) [inline tests/test_extract_child_mesh.cpp:113]: What's with all that commented out code? Where does it come from and why is it here?

**Review by daniel-zint — APPROVED** [review]
The functionality is there. Just clean it up and you're good to go.

Please be a more careful about when to use which integer data type. `int`,`long`, and `size_t` are not the same thing.

**Review by daniel-zint — APPROVED** [review]
lgtm

**Pull request merged**